3.0 - Microcode overhaul

mountainflaw opened the pull request:
The goal of this is to combine #789 and #764, while also moving microcode config from the makefile to config_graphics.h

Commit:
F3DZEX to object file

Commit:
Microcode selection to config_graphics.h

Commit:
Remove build options display.

Commit:
Microcode massacre

Commit:
Total HVQM death

Commit:
Remove S2DEX engine

Commit:
Use enumatered F3DEX selector

Commit:
ditto

Commit:
Fix F3DEX build

Commit:
Remove hvqm remnants

Commit:
fix mislabeled config safeguard entry

Commit:
Rename L3DZEX2

Commit:
Add preliminary F3DEX3 support

Commit:
.

Commit:
Simplify grucode pointers

Commit:
Decringify i guess?

Commit:
Remove gs2dex includes

Commit:
.

Commit:
fixed newlines in cfg

Commit:
Fix lighting direction for F3DEX3

Commit:
Add WIP F3DEX3 performance profiler

Commit:
wtf

Commit:
Fix config issues

Commit:
Move F3DEX3 CPU functionality to a dedicated file

Commit:
Readme credit

Commit:
Use new profiler controls

Commit:
Restore config

ThePerfectMario64 commented:
silly.

Commit:
update ex3 version + fix lvp error

Commit:
F3DEX3 update (real)

Commit:
Fix area.c viewport

Commit:
add buffer clear optimizations

Commit:
fix non ex3 viewport

someone2639 commented:
lgtm (just fix the merge conflicts)

Commit:
goddard fixes

Commit:
fix memset in game_init.c label

Commit:
ending screen viewport fix

Commit:
finish goddard for now

Commit:
Remove redundant check for lookat calculation

Commit:
Add gSPCameraWorld for ingame graphics

Commit:
Use G_LIGHTING_SPECULAR for reflective surfaces

Commit:
add define for ex3 memclear, more specular changes

Commit:
Add disable instant input config

Commit:
cfb clear coverage reset

Commit:
workaround for memclear cfb oddity

Commit:
Add specular shading to DDD

Commit:
fix framebuffer garbage on goddard load

Commit:
ditto

Commit:
ditto

Commit:
fix goddard renderer typo in cameraworld apply

Commit:
cleanup lookat code

mountainflaw committed:
Merge pull request #3 from mountainflaw/testingg

Testingg

mountainflaw commented:
Looks like a lot of cleanup needs to be done but what's left are just occlusion planes and possibly a basic lighting engine for the new point lights (should be mostly self contained and not require any large overhauls). I found a way to get around Git being funky with handling commit history on this branch.

mountainflaw committed:
Merge branch 'develop/3.0.0' into ucode_to_cfg

mountainflaw committed:
Update config.h

mountainflaw commented:
conflicts resolved

Commit:
ttc speculars

Commit:
Add simple F3DEX3 lighting engine

Commit:
Sort linked list before render to render closest to mario point lights first

Commit:
fix lightnum bug

Commit:
code cleanup

Commit:
fix ex2 build

Commit:
Update lighting engine info

Commit:
Remove F3DEX support

Commit:
use bcopy for struct data transfer in lighting engine

Commit:
delete debug stuff for lighting engine

Commit:
Remove RSP memclear

Commit:
lighting engine: stop early with no lights

Commit:
Add F3DEX3 lighting engine documentation

Commit:
Fix end credits

mountainflaw committed:
Merge pull request #4 from mountainflaw/testingg

Testingg

mountainflaw commented:
Fully tested. Ready for review.

arthurtilly commented on actors/metal_box/model.inc.c:
```diff
@@ -46,6 +46,9 @@ const Gfx metal_box_seg8_dl_08024B18[] = {
     gsSPLightColor(LIGHT_2, 0x7f7f7fff),
+#ifdef F3DEX_GBI_3
+    gsSPSetGeometryMode(G_LIGHTING_SPECULAR),
+#endif
```
nope

arthurtilly reviewed (changes requested):
nice try idiot

Commit:
remove actor specular

Commit:
level unspecularize

mountainflaw committed:
Merge pull request #5 from mountainflaw/testingg

Testingg

Commit:
despecularized segment2

mountainflaw committed:
Merge pull request #6 from mountainflaw/testingg

despecularized segment2

arthurtilly commented on include/config.h:
```diff
@@ -19,10 +19,4 @@
-// Local config - include a gitignore'd config file that's specific to just the user (if the file exists)
-#if __has_include("config/config_local.h")
-    #include "config/config_local.h"
-#endif
```
Why's this removed

arthurtilly commented on include/config/config_game.h:
```diff
@@ -100,3 +100,4 @@
  * The levelscript needs to have a MARIO_POS command for this to work.
  */
 #define START_LEVEL LEVEL_CASTLE_GROUNDS
+
```
arceveti

mountainflaw replied:
that's not an arceveti

arthurtilly replied:
its literally the only change in this file

mountainflaw replied:
oh i guess it did end up that way on your end. i try to make sure all source and header files i work with have a newline at the end so this is indeed unnecessary since i never touch it

mountainflaw replied:
resolved

arthurtilly commented on include/config/config_graphics.h:
```diff
@@ -25,7 +25,7 @@
  * Spawns floombas on the title screen.
  */
-#define INTRO_FLOOMBAS
+//#define INTRO_FLOOMBAS
```
whyd you change this...

arthurtilly commented on include/config/config_safeguards.h:
```diff
@@ -112,12 +150,20 @@
     #define DEBUG_ASSERTIONS
 #endif // DEBUG
 
-
```
arceveti

arthurtilly commented on levels/intro/script.c:
```diff
@@ -98,11 +99,12 @@ const LevelScript level_intro_mario_head_regular[] = {
-    BLACKOUT(/*active*/ FALSE),
-    LOAD_AREA(/*area*/ 1),
     SET_MENU_MUSIC(/*seq*/ SEQ_MENU_TITLE_SCREEN),
+    LOAD_AREA(/*area*/ 1),
```
what are these changes

mountainflaw replied:
fixes goddard framebuffer garbage

arthurtilly commented on src/game/game_init.c:
```diff
@@ -33,7 +35,7 @@
 
 // Emulators that the Instant Input patch should not be applied to
-#define INSTANT_INPUT_BLACKLIST (EMU_CONSOLE | EMU_WIIVC | EMU_ARES | EMU_SIMPLE64 | EMU_CEN64)
+#define INSTANT_INPUT_BLACKLIST (EMU_CONSOLE | EMU_WIIVC | EMU_ARES | EMU_SIMPLE64 | EMU_CEN64 | DISABLE_INSTANT_INPUT)
```
whats this

mountainflaw replied:
a new config option that forces instant input to be off when defined. made testing a lot easier.

arthurtilly replied:
it's not referenced anywhere except safeguards as far as i can tell, so it should be removed

mountainflaw replied:
https://github.com/HackerN64/HackerSM64/pull/816/files#diff-544ef2df343d5825e817af38eba0b3908b7d4f1e5d652dce777398032a4c023bR160

arthurtilly replied:
oh im blind. still not sure abt it. would rather it was in this file than config

mountainflaw replied:
ideally it would get added to something like disable_all in the future, instant input does pop up and cause problems frequently

arthurtilly commented on src/game/rendering_graph_node.c:
```diff
@@ -745,7 +759,6 @@ void geo_process_background(struct GraphNodeBackground *node) {
         Gfx *gfxStart = alloc_display_list(sizeof(Gfx) * 8);
 #endif
         Gfx *gfx = gfxStart;
-
```
arceveti

arthurtilly commented on src/game/screen_transition.c:
```diff
@@ -130,7 +133,6 @@ u16 calc_tex_transition_direction(struct WarpTransitionData *transData) {
  */
 s32 render_textured_transition(u8 transTime, struct WarpTransitionData *transData, s8 texID, s8 transTexType) {
     u16 texTransDir = calc_tex_transition_direction(transData);
-
```
more arecevetis

arthurtilly commented on src/game/screen_transition.c:
```diff
@@ -181,6 +183,7 @@ s32 render_textured_transition(u8 transTime, struct WarpTransitionData *transDat
 
         gDisplayListHead = tempGfxHead;
     }
+
```
arceveti

arthurtilly commented on src/game/screen_transition.c:
```diff
@@ -235,7 +238,6 @@ u8 set_transition_color_fade_alpha(s8 fadeType, u8 transTime) {
 
 s32 render_fade_transition_from_color(u8 transTime, struct WarpTransitionData *transData) {
     u8 alpha = set_transition_color_fade_alpha(COLOR_TRANS_FADE_FROM_COLOR, transTime);
-
```
arceveti

Commit:
restore config

Commit:
defuck spacing

Commit:
ditto

Commit:
game

Commit:
bring back the cringe

Commit:
cleanup goddard lookat

mountainflaw commented:
Total changelog with current state of this branch:

- Removed all microcodes except F3DZEX2, F3DEX3, audio, and of course, boot.
  - All supporting code for removed microcodes has been removed as well.
- Fixed Goddard framebuffer garbage being visible.
- Reduced gSPLookAt to one command per frame.
- Added specular lighting to Goddard screen, among other minor changes for F3DEX3.
- Added F3DEX3 debug profiler.
- Added instant input override.
- Added F3DEX3 lighting engine (enabled by default with F3DEX3), removed mention of wiseguy's in the README.md as this runs faster and outdoes it in every category while remaining vanilla compatible.
- Uses gSPMemSet on Z buffer for faster clear.

arthurtilly commented on include/config/config_graphics.h:
```diff
@@ -152,3 +152,41 @@
+ * Disables instant input.
+ * Mostly used for debugging purposes, or for future-proofing a release if concerned.
+ */
+//  #define DISABLE_INSTANT_INPUT
```
Move this into game_init.c and reverse it so it is #define INSTANT_INPUT

mountainflaw replied:
blah blah blah fixed

arthurtilly commented on src/game/mario_actions_cutscene.c:
```diff
@@ -37,12 +37,21 @@ static s16 sEndPeachAnimation;
+        { (SCREEN_WIDTH  * 2), (SCREEN_HEIGHT * 2), G_NEW_MAXZ / 2, 0 }
+    }
+};
+#else
```
was end cutscene viewport ever fixed?

mountainflaw replied:
yes

arthurtilly commented on src/goddard/renderer.c:
```diff
@@ -2260,13 +2249,25 @@ static void gd_dl_viewport(void) {
+    vp->vp.vtrans[2] = G_NEW_MAXZ / 2;
+#else
+    vp->vp.vtrans[2] = 0x1FF;
+#endif
```
these ifdefs here are a bit ugly. if original maxz was 0x3FE then this could be handled above the function and be cleaner? but idk how it works

mountainflaw replied:
yeah they can be combined i guess

arthurtilly reviewed (changes requested):
shouldve done this as one review but oh well

gheskett commented:
Abstaining from review until Arthur's comments have been resolved.

Commit:
Fix lighting engine bug

Commit:
beta

mountainflaw commented:
beta

Commit:
remove instant input global def

mountainflaw commented:
ready!!!!

arthurtilly commented:
Closed due to the "no beta" rule.

mountainflaw commented:
this is so sad

BeverlyBean commented:
https://github.com/user-attachments/assets/d05c0713-e983-4779-8486-1d7f685d60ba

Commit:
reduce ifdefs in sm64.ld

aglab2 commented on lib/PR/f3dex3/gspF3DEX3_BrW_PA.o:
Will be a blocker for eabi in Hacker, please use .s files + .code/.data files

mountainflaw replied:
fixed

aglab2 commented on src/game/game_init.c:
```diff
@@ -142,36 +147,26 @@ const Gfx init_rsp[] = {
         return;
+    }
+
+    gDPSetColorImage(tempGfxHead++, G_IM_FMT_RGBA, G_IM_SIZ_16b, SCREEN_WIDTH, gPhysicalZBuffer);
```
Arceveti without any changes? I'd expect a memset here.

gheskett replied:
I feel like it doesn't make sense to use a CPU operation to clear it since it isn't buffered.

mountainflaw replied:
he is referring to rsp memset. i couldn't get it to work with the game and ended up fixing inconsistent formatting here when i did try to use it

BeverlyBean commented:
if ex3 is still subject to changes then shouldn't this go into a `develop` branch

Commit:
goddard clean

Commit:
fixed it for you aglab

Commit:
a

aglab2 commented on Makefile:
```diff
@@ -421,7 +376,6 @@ O_FILES := $(foreach file,$(C_FILES),$(BUILD_DIR)/$(file:.c=.o)) \
            $(foreach file,$(CPP_FILES),$(BUILD_DIR)/$(file:.cpp=.o)) \
            $(foreach file,$(S_FILES),$(BUILD_DIR)/$(file:.s=.o)) \
            $(foreach file,$(GENERATED_C_FILES),$(file:.c=.o)) \
```
Stray `\`

aglab2 commented on include/config/config_graphics.h:
```diff
@@ -152,3 +152,32 @@
+ * NOTE: Game looks visually identical if not setup.
+ */
+
+#define F3DEX3_LIGHTING_ENGINE
```
Is this OK to have this as a default especially when EX2 is default?

gheskett replied:
I don't know the answer to this either, can you add a clarification comment suggesting EX2 usage disables this by default?

aglab2 commented on include/ique/PR/mbi.h:
```diff
@@ -79,8 +79,6 @@
 #define	M_GFXTASK	1
 #define	M_AUDTASK	2
 #define	M_VIDTASK	3
-#define M_HVQTASK	6
```
Were those task ids added by HackerSM64 or in SDK by default? I'd rather not patch original files if so.

mountainflaw replied:
fixed

aglab2 commented on src/game/f3dex3.c:
```diff
@@ -0,0 +1,388 @@
+#ifdef DEBUG_F3DEX3_PROFILER
+u32 gF3DEX3ProfilerPage = 0;
+
+volatile F3DEX3YieldDataFooter gRSPProfilingResults;
```
No point in `volatile` here. What must be marked as `volatile` is pointer `volatile F3DEX3YieldDataFooter* footer` in `extract_f3dex3_profiler_data`

mountainflaw replied:
that's how sauraen has it and trying to do that gave me an error

aglab2 replied:
fine for now

aglab2 commented on src/game/f3dex3.c:
```diff
@@ -0,0 +1,388 @@
+        return 0.0f;
+    }
+
+    register f32 distX = ABS(gMarioState->pos[0] - (f32)x);
```
Remove `register`, it is bogus in modern C

mountainflaw replied:
fixed

aglab2 commented on src/game/f3dex3.c:
```diff
@@ -0,0 +1,388 @@
+    sGlobalAmbientLight->l.col[1] = g;
+    sGlobalAmbientLight->l.col[2] = b;
+
+    bcopy(&sGlobalAmbientLight->l.col[0], &sGlobalAmbientLight->l.colc[0], sizeof(u8) * 3);
```
Seems like an overkill for something that can be done with a simple `*(u32*)&sGlobalAmbientLight->l.col[0]=*(u32*)&sGlobalAmbientLight->l.colc[0]`

mountainflaw replied:
fixd

aglab2 commented on src/game/game_init.c:
```diff
@@ -142,36 +147,26 @@ const Gfx init_rsp[] = {
         return;
+    }
+
+    gDPSetColorImage(tempGfxHead++, G_IM_FMT_RGBA, G_IM_SIZ_16b, SCREEN_WIDTH, gPhysicalZBuffer);
```
Should be moved above `resetZB`.

mountainflaw replied:
why waste the rdp command on that if it's false

aglab2 commented on src/game/rendering_graph_node.c:
```diff
@@ -604,6 +609,15 @@ void geo_process_camera(struct GraphNodeCamera *node) {
 #endif
     gSPMatrix(gDisplayListHead++, VIRTUAL_TO_PHYSICAL(viewMtx), G_MTX_PROJECTION | G_MTX_MUL | G_MTX_NOPUSH);
+
+#ifdef F3DEX_GBI_3
```
Only needed when `F3DEX3_LIGHTING_ENGINE` is enabled?

mountainflaw replied:
yes that is all unique to ex3

aglab2 commented on src/goddard/renderer.c:
```diff
@@ -1850,24 +1864,8 @@ void gd_dl_lookat(struct ObjCamera *cam, f32 arg1, f32 arg2, f32 arg3, f32 arg4,
     lookat->l[1].l.dir[1] = LOOKAT_PACK(cam->unkE8[1][1]);
     lookat->l[1].l.dir[2] = LOOKAT_PACK(cam->unkE8[2][1]);
-
-    lookat->l[0].l.col[0] = 0;
```
Still needed when EX2 is enabled

mountainflaw replied:
does not affect ex2

Commit:
fix warning with profiler enabled

Commit:
opsie

Commit:
fix nits

Commit:
remove the most useless fucking folder ever

mountainflaw committed:
Merge branch 'develop/3.0.0' into ucode_to_cfg

Commit:
makefile cleanup

Commit:
opsie2

mountainflaw commented:
rundown to people just now reviewing:
- removes s2dex engine + microcode
- removes hvqm microcode + code
- removes ique include directory because it's 1:1 with the n64 one
- removes every microcode except f3dzex2 non, and add f3dex3 + the cpu code
- microcode is now a config in include/config_graphics.h
- adds an f3dex3-exclusive lighting engine
- updates the face demo to use specular lighting
- removes armips because the output for it was not being used anyways
- removes various other leftovers from ultrasm64's microcode setup that were not being used
- reintroduces rsp.s from decomp, except with much better formatting to allow for easier abi changing

aglab2 reviewed (approved):
Looks ok to me

Commit:
more profiler changes from zelda, add flush

Commit:
Fix tools makefile

mountainflaw commented:
it is time to review #816

aglab2 commented on lib/rsp.s:
```diff
@@ -0,0 +1,114 @@
+    glabel gspF3DEX3_BrWTextEnd
+    #ifdef DEBUG_F3DEX3_PROFILER
+        .balign 16
+        glabel gspF3DEX3_BrW_PATextStart
```
Will these microcodes get removed in build time if unused? Does `-fdata-sections` work with asm files? If not, let's explicitly place those files in different sections

mountainflaw replied:
They are only used if the profiler is enabled. Disabling the profile causes a rebuild due to being a config change.

mountainflaw replied:
I do not know if `-fdata-sections` works on gas though

gheskett commented on README.md:
```diff
@@ -31,11 +31,20 @@ This is a fork of the ultrasm64 repo by CrashOveride which includes the followin
 
-**Lighting Engine by Wiseguy**
-- Lighting Engine is available on a separate branch ([base/lighting-engine](https://github.com/Reonu/HackerSM64/tree/base/lighting-engine)). Instructions on how to use it are in the readme of that branch.
+**Lighting Engine by red**
+- A new lighting engine for the F3DEX3 is included with the repository. Enable `F3DEX3_LIGHTING_ENGINE` within `config_graphics.h` to use.
+- Improvements to the older lighting engine by wiseguy:
+    - Support for variable number of directional lights at once.
+    - Support for vanilla actors.
+        - Limits you to up to 7 extra lights when setup this way (`gLightNumBase = NUMLIGHTS_3`), but for scenes tailored specifically for a lighting engine (`gLightNumBase = NUMLIGHTS_1`), you can have up to 9 variable lights.
+        - Ambient light does not count towards the light limit.
+        - F3DEX3 fixes point lights with kc < 8, allowing for more expressive lighting.
+        - F3DEX3 also fixes point light behavior when close to triangles.
+- See `src/game/f3dex3.c` for more details.
```
Is old LE actually fully obsoleted by f3dex3 (aside from emulator compatibility)? If not, I would leave mention of Wiseguy LE in here (though honestly emu compat by itself kinda justifies it).

mountainflaw replied:
Yes, it can do everything that old LE can do and has more features.

gheskett commented on include/config/config_debug.h:
```diff
@@ -94,3 +94,10 @@
+ * Enable F3DEX3 profiling tool. Hold L + C Down to display and cycle through profiler views.
+ * Using pages 2, 3, and 4 can disable certain F3DEX3 features. Read the F3DEX3 README for more information.
+ */
+
```
Inconsistent config whitespace

gheskett commented on include/config/config_safeguards.h:
```diff
@@ -49,6 +49,33 @@
+    #elif F3DEX_VERSION == 3
+        #define F3DEX_GBI_3
+    #else
+        #error "Invalid microcode selection. Valid options are 1-3 inclusive."
```
F3DEX1 is a lie

gheskett commented on include/config/config_graphics.h:
```diff
@@ -152,3 +152,32 @@
+ * See README.md and src/game/f3dex3.c for more details.
+ * NOTE: Game looks visually identical if not setup.
+ */
+
```
Inconsistent config whitespace

gheskett commented on include/config/config_graphics.h:
```diff
@@ -152,3 +152,32 @@
+ * This feature is not available for use with F3DEX3.
+ */
+
+// #define ENABLE_LINE_UCODE
```
Hmm, is there any legitimate reason to not just ship this by default with EX2?

mountainflaw replied:
I mean it would be an extra 8 KB of RAM wasted

gheskett replied:
But is 8kb really worth an entire config define? That's 2 textures.

arthurtilly replied:
why is this not just in the makefile isnt that where it was before

gheskett commented on include/config/config_safeguards.h:
```diff
@@ -112,6 +139,15 @@
     #define UNLOCK_ALL
 #endif // COMPLETE_SAVE_FILE
 
+#ifdef DEBUG_F3DEX3_PROFILER
```
This should also be undef'd whenever `DISABLE_ALL` is enabled

gheskett commented on levels/intro/script.c:
```diff
@@ -98,11 +99,12 @@ const LevelScript level_intro_mario_head_regular[] = {
+    LOAD_AREA(/*area*/ 1),
     TRANSITION(/*transType*/ WARP_TRANSITION_FADE_FROM_STAR, /*time*/ 20, /*color*/ 0x00, 0x00, 0x00),
-    SLEEP(/*frames*/ 20),
+    SLEEP(/*frames*/ 2),
```
Why was this changed to 2?

gheskett commented on src/boot/main.c:
```diff
@@ -22,6 +22,9 @@
 #include "game/puppyprint.h"
 #include "game/profiling.h"
 #include "game/emutest.h"
+#ifdef F3DEX_GBI_3
+#include "game/f3dex3.h"
+#endif
```
I would move the ifdef to the actual f3dex3 header rather than for each individual f3dex3 include

gheskett commented on src/game/area.c:
```diff
@@ -26,8 +27,8 @@
 #include "profiling.h"
-#ifdef S2DEX_TEXT_ENGINE
-#include "s2d_engine/init.h"
+#ifdef F3DEX_GBI_3
+#include "f3dex3.h"
 #endif
```
DItto

gheskett commented on src/game/area.c:
```diff
@@ -374,6 +384,10 @@ void play_transition_after_delay(s16 transType, s16 time, u8 red, u8 green, u8 b
     play_transition(transType, time, red, green, blue);
 }
 
+/**
+ * Draws the F3DEX3 profiler.
+ */
```
`render_game` does _much_ more than just draw a F3DEX3 profiler...

mountainflaw replied:
so true bestie

gheskett commented on src/game/area.c:
```diff
@@ -410,6 +424,10 @@ void render_game(void) {
         } else
             gDPSetScissor(gDisplayListHead++, G_SC_NON_INTERLACE, 0, gBorderHeight, SCREEN_WIDTH,
                           SCREEN_HEIGHT - gBorderHeight);
+        
+        #ifdef DEBUG_F3DEX3_PROFILER
+            draw_f3dex3_profiler();
+        #endif
```
Should we move this to outside the if/else block so that it always renders?

Also nitpick: Unindent preprocessor stuff here to syntactically match the rest of the function/game

gheskett commented on src/game/f3dex3.c:
```diff
@@ -0,0 +1,388 @@
+#include <ultra64.h>
+
+#ifdef F3DEX_GBI_3
+#include "game_init.h"
+#ifdef DEBUG_F3DEX3_PROFILER
+#include "buffers/buffers.h"
+#include "puppyprint.h"
+#include "fasttext.h"
+#endif
```
Nit: Space these out a bit to make them easier to visually parse, and/or add comments to the `endif`s

gheskett commented on src/game/area.c:
```diff
@@ -410,6 +424,10 @@ void render_game(void) {
         } else
             gDPSetScissor(gDisplayListHead++, G_SC_NON_INTERLACE, 0, gBorderHeight, SCREEN_WIDTH,
                           SCREEN_HEIGHT - gBorderHeight);
```
Unrelated, but now that I'm seeing it, please can this `else` match the `if` in terms of bracket usage (or lack thereof)?

gheskett commented on src/game/f3dex3.c:
```diff
@@ -0,0 +1,388 @@
+    *(u32*)&sGlobalAmbientLight->l.colc[0] = *(u32*)&sGlobalAmbientLight->l.col[0];
+}
+
+static f32 calculate_distance_from_mario(s16 x, s16 y, s16 z) {
+    if (gMarioState->marioObj == NULL) { // No distance calc on Mario-less levels
+        return 0.0f;
+    }
+
+    f32 distX = ABS(gMarioState->pos[0] - (f32)x);
+    f32 distY = ABS(gMarioState->pos[1] - (f32)y);
+    f32 distZ = ABS(gMarioState->pos[2] - (f32)z);
+
+    return sqr(distX + distY + distZ);
+}
```
Does it make more or less sense to use Mario state instead of Mario object here?

mountainflaw replied:
I think it would just be preference

gheskett commented on src/game/game_init.c:
```diff
@@ -19,22 +21,25 @@
 #ifdef SRAM
 #include "sram.h"
 #endif
+#ifdef F3DEX_GBI_3
+#include "f3dex3.h"
+#endif
```
remove ifdef

gheskett commented on src/game/rendering_graph_node.c:
```diff
@@ -19,8 +21,13 @@
 #include "color_presets.h"
 #include "emutest.h"
 
+#ifdef F3DEX3_LIGHTING_ENGINE
+#include "f3dex3.h"
+#endif
```
remove ifdef

gheskett commented on src/game/rendering_graph_node.c:
```diff
@@ -348,9 +354,6 @@ void geo_process_master_list_sub(struct GraphNodeMasterList *node) {
  * render modes of layers.
  */
 void geo_append_display_list(void *displayList, s32 layer) {
-#ifdef F3DEX_GBI_2
-    gSPLookAt(gDisplayListHead++, gCurLookAt);
-#endif
```
What was this doing before?

mountainflaw replied:
It's redundant.

gheskett commented on src/goddard/renderer.c:
```diff
@@ -1982,22 +1962,22 @@ void func_801A0038(void) {
-                sTriangleBuf[i][1] - sVertexBufStartIndex,
-                sTriangleBuf[i][2] - sVertexBufStartIndex,
-                0);
+            if (sTriangleBufCount - i > 1) {
+                gSP2Triangles(&sCurrentGdDl->gfx[sCurrentGdDl->curGfxIdx++], sTriangleBuf[i][0] - sVertexBufStartIndex, sTriangleBuf[i][1] - sVertexBufStartIndex, sTriangleBuf[i][2] - sVertexBufStartIndex, 0,
+                                        sTriangleBuf[i + 1][0] - sVertexBufStartIndex, sTriangleBuf[i + 1][1] - sVertexBufStartIndex, sTriangleBuf[i + 1][2] - sVertexBufStartIndex, 0);
+                                        i++;
+            } else {
```
This is a complete indentation mess (at least in the GitHub view)

Commit:
github conversations go brrrr

mountainflaw commented:
fixed
